File: ddd-application/src/main/java/victor/training/ddd/agile/entity/Product.java
package victor.training.ddd.agile.entity;

import lombok.Data;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import java.util.ArrayList;
import java.util.List;


@Data
@Entity
public class Product {
   @Id
   @GeneratedValue
   private Long id;
   private int currentIteration = 0;
   private int currentVersion = 0;
   private String code;
   private String name;

   private String ownerEmail;
   private String ownerName;
   private String ownerPhone;

   private String teamMailingList;

   @OneToMany(mappedBy = "product")
   private List<Sprint> sprints = new ArrayList<>();

   @OneToMany(mappedBy = "product")
   private List<Release> releases = new ArrayList<>();

   public int incrementAndGetIteration() {
      currentIteration++;
      return currentIteration;
   }

   public int incrementAndGetVersion() {
      currentVersion++;
      return currentVersion;
   }
}

File: ddd-application/src/main/java/victor/training/ddd/agile/entity/Product.kt
package victor.training.ddd.agile.entity

import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.OneToMany

@Entity
class Product(
    val code: String,
    val name: String,
    val teamMailingList: String,
    private var currentIteration: Int = 0,
    private var currentVersion: Int = 0,
    @OneToMany(mappedBy = "product")
    val releases: MutableList<Release> = ArrayList(),
    @Id
    @GeneratedValue
    var id: Long? = null
) {

    fun incrementAndGetIteration(): Int {
        currentIteration++
        return currentIteration
    }

    fun incrementAndGetVersion(): Int {
        currentVersion++
        return currentVersion
    }
}